File: code/backend-java/src/main/java/techwatch/ai/orchestrator/entity/Article.java
package techwatch.ai.orchestrator.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Represents an article retrieved via an RSS feed or a website.
 */
@Entity
@Table(name = "articles")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Article {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    // The URL is unique to avoid duplicates during successive scans
    @Column(unique = true, nullable = false, length = 1024)
    private String link;

    private String sourceName;

    private LocalDateTime pubDate;

    // Field for the full content once scraped by the Python service
    @Column(columnDefinition = "TEXT")
    private String fullContent;

    // Audit: allows knowing when the entry was created in the database
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}

File: code/backend-python/app/models/article.py
import uuid
from sqlalchemy import Column, String, DateTime, Text
from sqlalchemy.dialects.postgresql import UUID
from pgvector.sqlalchemy import Vector
from app.core.database import Base

class Article(Base):
    """
    SQLAlchemy model representing the 'articles' table.
    Shared with the Spring Boot orchestrator and includes pgvector support for semantic search.
    """
    __tablename__ = "articles"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    title = Column(String, nullable=False)
    description = Column(Text)
    link = Column(String, unique=True, nullable=False)
    source_name = Column(String, name="source_name")
    pub_date = Column(DateTime, name="pub_date")
    full_content = Column(Text, name="full_content")
    created_at = Column(DateTime, name="created_at")

    # Vector embeddings (768 dimensions optimized for the nomic-embed-text model)
    embedding = Column(Vector(768), name="embedding")